pkg/wait/wait_for_ready.go

```diff
@@ -0,0 +1,158 @@
+type waitForReadyConfig struct {
+	watchFunc           WatchFunc
+	conditionsExtractor ConditionsExtractor
+	kind                string
+}
+
+// Interface used for waiting of a resource of a given name to reach a definitive
+// state in its "Ready" condition.
+type WaitForReady interface {
+
+	// Wait on resource the resource with this name until a given timeout
+	// and write status out on writer
+	Wait(name string, timeout int, out io.Writer) error
+}
+
+// Create watch which is used when waiting for Ready condition
+type WatchFunc func(opts v1.ListOptions) (watch.Interface, error)
+
+// Extract conditions from a runtime object
+type ConditionsExtractor func(obj runtime.Object) (apis.Conditions, error)
+
+// Constructor with resource type specific configuration
+func NewWaitForReady(kind string, watchFunc WatchFunc, extractor ConditionsExtractor) WaitForReady {
+	return &waitForReadyConfig{
+		kind:                kind,
+		watchFunc:           watchFunc,
+		conditionsExtractor: extractor,
+	}
+}
+
+// Wait until a resource enters condition of type "Ready" to "False" or "True".
+// `watchFunc` creates the actual watch, `kind` is the type what your are watching for
+// (e.g. "service"), `timeout` is a timeout after which the watch should be cancelled if no
+// target state has been entered yet and `out` is used for printing out status messages
+func (w *waitForReadyConfig) Wait(name string, timeout int, out io.Writer) error {
+	opts := v1.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector("metadata.name", name).String(),
+	}
+	addWatchTimeout(&opts, timeout)
+
+	watcher, err := w.watchFunc(opts)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(out, "Waiting for %s '%s' to become ready ... ", w.kind, name)
+	flush(out)
+
+	if err := w.waitForReadyCondition(watcher, name, timeout); err != nil {
+		fmt.Fprintln(out)
+		return err
+	}
+	fmt.Fprintln(out, "OK")
+	return nil
+}
+
+func addWatchTimeout(opts *v1.ListOptions, timeout int) {
+	if timeout == 0 {
+		return
+	}
+	// Wait for service to enter 'Ready' state, with a timeout of which is slightly larger than
+	// the provided timeout. We have our own timeout which fires after "timeout" seconds
+	// and stops the watch
+	timeOutWatch := int64(timeout + 30)
+	opts.TimeoutSeconds = &timeOutWatch
+}
+
+// Duck type for writers having a flush
+type flusher interface {
+	Flush() error
+}
+
+func flush(out io.Writer) {
+	if flusher, ok := out.(flusher); ok {
+		flusher.Flush()
+	}
+}
+
+func (w *waitForReadyConfig) waitForReadyCondition(watcher watch.Interface, name string, timeout int) error {
+	defer watcher.Stop()
+	for {
+		select {
+		case <-time.After(time.Duration(timeout) * time.Second):
+			return fmt.Errorf("timeout: %s '%s' not ready after %d seconds", w.kind, name, timeout)
+		case event, ok := <-watcher.ResultChan():
+			if !ok || event.Object == nil {
+				return fmt.Errorf("timeout while waiting for %s '%s' to become ready", w.kind, name)
```

Comment:
this case is not an error but rather it's necessary to retry watching. k8s will expire watchers after a few minutes consistently. see example here: https://github.com/k14s/kapp/blob/0c91e8a2208ff9a8f9d92f8c01d20dc9dd21256b/pkg/kapp/resources/pod_watcher.go#L64

Reply:
I see the point. Let's change that from an error to an informal message and add a retry mechanism later (as mentioned in https://github.com/knative/client/pull/156#discussion_r291419076), then with a real timeout.

Reply:
I added now retry mechanism when the watch closes. Not really sure whether this will happen often as I use a timeout for the list watch which is 30s longer than the user provided timeout. But as a watch can stop anytime anyway, its indeed saver to retry.